src/main.rs

```diff
@@ -56,52 +68,108 @@ fn main() {
         return;
     }
 
-    let optimization_level = if env::args().any(|arg| arg == "--release") {
-        String::from("release")
-    } else {
-        String::from("debug")
-    };
-
-    // Skip `cargo 3ds`
-    let mut args = env::args().skip(2);
-
     // Get the command and collect the remaining arguments
-    let command = args.next();
-    let args: Vec<String> = args.collect();
-    let args: Vec<&str> = args.iter().map(String::as_str).collect();
-
-    let (command, must_link) = match command.as_deref() {
-        Some("link") => ("build", true),
-        Some(command) => (command, false),
-        None => {
-            print_usage(&mut io::stderr());
-            process::exit(2)
-        }
-    };
+    let cargo_command = CargoCommand::from_args().unwrap_or_else(|| {
+        print_usage(&mut io::stderr());
+        process::exit(2)
+    });
 
     eprintln!("Running Cargo");
-    build_elf(command, &args);
+    let (status, messages) = cargo_command.build_elf();
+    if !status.success() {
+        process::exit(status.code().unwrap_or(1));
+    }
 
-    if command != "build" && !must_link {
-        // We only do more work if it's a build or build + 3dslink operation
+    if !cargo_command.should_build_3dsx() {
         return;
     }
 
     eprintln!("Getting metadata");
-    let app_conf = get_metadata(&args, &optimization_level);
+    let app_conf = get_metadata(&messages);
 
-    eprintln!("Building smdh");
+    eprintln!("Building smdh {:?}", app_conf.path_3dsx());
     build_smdh(&app_conf);
 
-    eprintln!("Building 3dsx");
+    eprintln!("Building 3dsx {:?}", app_conf.path_3dsx());
     build_3dsx(&app_conf);
 
-    if must_link {
+    if cargo_command.should_link {
         eprintln!("Running 3dslink");
         link(&app_conf);
     }
 }
 
+struct CargoCommand {
+    command: String,
+    should_link: bool,
+    args: Vec<String>,
+}
+
+impl CargoCommand {
+    fn from_args() -> Option<Self> {
+        // Skip `cargo 3ds`. `cargo-3ds` isn't supported for now
+        let mut args = env::args().skip(2);
+
+        let command = args.next()?;
+        let mut remaining_args: Vec<String> = args.collect();
+
+        let (command, should_link) = match command.as_str() {
+            "link" => ("build".into(), true),
```

Comment:
nit: avoid using `into` since it's not clear what it is actually converting into most of the time:
```suggestion
            "link" => ("build".to_string(), true),
```
(or `to_owned`)

Edit: there's a few more places like this.